crates/provider/src/heart.rs

```diff
@@ -15,9 +17,179 @@ use tokio::{
     sync::{mpsc, oneshot, watch},
 };
 
-/// A configuration object for watching for transaction confirmation.
-#[derive(Debug, Clone, Copy)]
-pub struct WatchConfig {
+/// A builder for configuring a pending transaction watcher.
+///
+/// # Examples
+///
+/// Send and wait for a transaction to be confirmed 2 times, with a timeout of 60 seconds:
+///
+/// ```no_run
+/// # async fn example<N: alloy_network::Network>(provider: impl alloy_provider::Provider<N>, tx: N::TransactionRequest) -> Result<(), Box<dyn std::error::Error>> {
+/// // Send a transaction, and configure the pending transaction.
+/// let builder = provider.send_transaction(tx)
+///     .await?
+///     .with_confirmations(2)
+///     .with_timeout(Some(std::time::Duration::from_secs(60)));
+/// // Register the pending transaction with the provider.
+/// let pending_transaction = builder.register().await?;
+/// // Wait for the transaction to be confirmed 2 times.
+/// let tx_hash = pending_transaction.await?;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// This can also be more concisely written using `watch`:
+/// ```no_run
+/// # async fn example<N: alloy_network::Network>(provider: impl alloy_provider::Provider<N>, tx: N::TransactionRequest) -> Result<(), Box<dyn std::error::Error>> {
+/// let tx_hash = provider.send_transaction(tx)
+///     .await?
+///     .with_confirmations(2)
+///     .with_timeout(Some(std::time::Duration::from_secs(60)))
+///     .watch()
+///     .await?;
```

Comment:
OK I am fine with this API -- let's open issue for potentially auto-registering if possible, but we can move forward.

Reply:
Not sure what this means